### ISSUE TITLE
Smarter formatting with object destructuring

### ISSUE THREAD

**Issue by RyanZim** [prettier/prettier#5988]
**Prettier 1.16.4**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAExMDMIQA0mARgIYBOSpZAXgIJQAmAyhALZwwAWAllAHNiARwCuADwCSaADLQBmAL6YAvJjQc4AIV4CAYqKhgYvaADkynABQAdEPgh2AlLZDkKzgNwhCICAAcTdGRQSgoIAHcABUoENGQQMgA3CF4mHzcKMjAAay4Wf2z+AWQYClE4X24YdgAbAHU+eDRCsDgWON4TJK6ATwSwNHjffjQ4ChgorIF2MmRcMlqx3wArNHEtLNz8yzgZfjh5xeWQNfEWYtq4AEVRCHgjpcqQQooxigTyEl7a6Az-Cj8GD1NI8ZAADgADL4ARAxvUsv4EgC4O8kodfBQ4GJeFipmQZnMkAsnr4xuxeKVys80JcbncHsTjs8YGQSCCmGCkAAmXxlMi8WrFADCHFmCSg0AxIFEYwAKmz4kynopFEA)
```sh
--parser babylon
```

This is just stylistic preference, but I'm wondering if most people wouldn't prefer better styling than what prettier does currently.

**Input:**
We have this big call, with some complicated destructuring
```jsx
const { foo, bar: bazAndSomething, quxIsLong } = someBigFunctionName("foo")("bar");
```

**Output:**
Prettier breaks up a single parameter function call, and keeps the complicated destructuring on one line
```jsx
const { foo, bar: bazAndSomething, quxIsLong } = someBigFunctionName("foo")(
  "bar"
);

```

**Expected behavior:**
This seems like a far more logical way to format the code:
```jsx
const {
  foo,
  bar: bazAndSomething,
  quxIsLong
} = someBigFunctionName("foo")("bar");
```

Basically, if we have a line that needs broken up, and it's an object destructuring assignment (especially if there's reassignment [i.e. `bar: bazAndSomething`]), prefer breaking the destructuring on to multiple lines instead of breaking up the function call.

**Comment by duailibe** [comment]
@RyanZim 

Agreed. We have a lot of these problems that two places can break and one would be better than the other. It's a hard problem to solve though.

I think I might have an idea, but we welcome anyone who wants to take a crack at it.

**Comment by bbugh** [comment]
This looks like the same issue as https://github.com/prettier/prettier/issues/3227

**Comment by RyanZim** [comment]
Closing as a dupe of https://github.com/prettier/prettier/issues/3227.
